identity-mdoc/src/main/java/com/android/identity/mdoc/mso/MobileSecurityObjectGenerator.kt

```diff
@@ -0,0 +1,276 @@
+     * later than the validFrom element.
+     * @param expectedUpdate Optional: if provided, represents the timestamp at which the issuing
+     * authority infrastructure expects to re-sign the MSO, else, null
+     * @return The `MobileSecurityObjectGenerator`.
+     * @throws IllegalArgumentException if the times are do not meet the constraints.
+     */
+    fun setValidityInfo(
+        signed: Timestamp,
+        validFrom: Timestamp, validUntil: Timestamp,
+        expectedUpdate: Timestamp?
+    ) = apply {
+        // 18013-5 Section 9.1.2.4 says "The timestamp of validFrom shall be equal or later than
+        // the signed element."
+        require(validFrom.toEpochMilli() >= signed.toEpochMilli()) {
+            "The validFrom timestamp should be equal or later than the signed timestamp"
+        }
+
+        // 18013-5 Section 9.1.2.4 says "The validUntil element contains the timestamp after which the
+        // MSO is no longer valid. The value of the timestamp shall be later than the validFrom element."
+        require(validUntil.toEpochMilli() > validFrom.toEpochMilli()) {
+            "The validUntil timestamp should be later than the validFrom timestamp"
+        }
+        mSigned = signed
+        mValidFrom = validFrom
+        mValidUntil = validUntil
+        mExpectedUpdate = expectedUpdate
+    }
+
+    private fun generateDeviceKeyBuilder(): CborBuilder {
+        val deviceKeyMapBuilder = CborMap.builder()
+        deviceKeyMapBuilder.put("deviceKey", mDeviceKey.toCoseKey(mapOf()).toDataItem)
+        if (!mAuthorizedNameSpaces.isEmpty() or !mAuthorizedDataElements.isEmpty()) {
+            val keyAuthMapBuilder = deviceKeyMapBuilder.putMap("keyAuthorizations")
+            if (!mAuthorizedNameSpaces.isEmpty()) {
+                val authNameSpacesArrayBuilder = keyAuthMapBuilder.putArray("nameSpaces")
+                for (namespace in mAuthorizedNameSpaces) {
+                    authNameSpacesArrayBuilder.add(namespace)
+                }
+                authNameSpacesArrayBuilder.end()
+            }
+            if (!mAuthorizedDataElements.isEmpty()) {
+                val authDataElemOuter = keyAuthMapBuilder.putMap("dataElements")
+                for (namespace in mAuthorizedDataElements.keys) {
+                    val authDataElemInner = authDataElemOuter.putArray(namespace)
+                    for (dataElemIdentifier in mAuthorizedDataElements[namespace]!!) {
+                        authDataElemInner.add(dataElemIdentifier)
+                    }
+                    authDataElemInner.end()
+                }
+                authDataElemOuter.end()
+            }
+            keyAuthMapBuilder.end()
+        }
+        if (!mKeyInfo.isEmpty()) {
+            val keyInfoMapBuilder = deviceKeyMapBuilder.putMap("keyInfo")
+            for (label in mKeyInfo.keys) {
+                keyInfoMapBuilder.put(label, mKeyInfo[label]!!)
+            }
```

Comment:
you can get both key and value of a map in 1 go 
(rather than getting the key "label" from the map just so you can get the bytes by accessing the map `mKeyInfo[label]`)

```
            for ((label, bytes) in mKeyInfo) {
                keyInfoMapBuilder.put(label, bytes)
            }
```